addons/mail/static/src/core/common/message_search_hook.js

```diff
@@ -0,0 +1,90 @@
+/* @odoo-module */
+
+import { useState, markup, onWillUnmount } from "@odoo/owl";
+import { useService } from "@web/core/utils/hooks";
+
+/**
+ * @param {string} searchTerm
+ * @param {string} target
+ */
+export function searchHighlight(searchTerm, target) {
+    if (!searchTerm) {
+        return target;
+    }
+    const htmlDoc = new DOMParser().parseFromString(target, "text/html");
+    const matchs = htmlDoc.evaluate(
+        `//*[text()[contains(translate(., '${searchTerm.toUpperCase()}', '${searchTerm.toLowerCase()}'), '${searchTerm.toLowerCase()}')]]`,
+        htmlDoc,
+        null,
+        XPathResult.ORDERED_NODE_SNAPSHOT_TYPE
+    );
+    for (let i = 0; i < matchs.snapshotLength; i++) {
+        const element = matchs.snapshotItem(i);
+        const newNode = [];
+        for (const node of element.childNodes) {
+            if (node.nodeType === Node.TEXT_NODE) {
+                const span = htmlDoc.createElement("span");
+                span.innerHTML = node.textContent.replaceAll(
+                    new RegExp(`(${searchTerm})`, "gi"),
+                    '<span class="bg-warning">$&</span>'
+                );
+                newNode.push(span);
+            } else {
+                newNode.push(node);
+            }
+        }
+        element.replaceChildren(...newNode);
+    }
+    return markup(htmlDoc.body.innerHTML);
+}
+
+/**
+ * @param {import('@mail/core/common/thread_model').Thread} thread
+ */
+export function useMessageSearch(thread) {
+    /** @type {import('@mail/core/common/thread_service').ThreadService} */
+    const threadService = useService("mail.thread");
+
+    function clear() {
+        state.messages = [];
+        state.searched = false;
+        state.searchTerm = undefined;
+    }
+
+    function search() {
+        if (state.searchTerm) {
+            const before = state.messages
+                ? Math.min(...state.messages.map((message) => message.id))
```

Comment:
why are we only searching among messages that are older than the oldest previously fetched? it looks like an optimisation in the case where the new `searchTerm` is a subset of the old `searchTerm`.

But when the `searchTerm` completely changes (which can happen without triggering a `search()` without `searchTerm` and therefore not clearing the `state.message[]`):
* if the match is older than the oldest message in `state.messages[]` it will find it and append it to the current results (maybe strange behavior).
* if the match is younger than the oldest message in `state.messages[]`, nothing will be found.

It looks like there should be a call to `clear()` if `searchTerm` does not contain the "old" `searchTerm`.